rspace/src/main/scala/coop/rchain/rspace/nextgenrspace/history/RootRepository.scala

```diff
@@ -0,0 +1,38 @@
+package coop.rchain.rspace.nextgenrspace.history
+
+import cats.Applicative
+import cats.implicits._
+import cats.effect.Sync
+import coop.rchain.rspace.Blake2b256Hash
+
+class RootRepository[F[_]: Sync](
+    rootsStore: RootsStore[F]
+) {
+  def commit(root: Blake2b256Hash): F[Unit] =
+    for {
+      _ <- rootsStore.recordRoot(root)
+    } yield ()
+
+  def currentRoot(): F[Blake2b256Hash] =
+    for {
+      maybeRoot <- rootsStore.currentRoot()
+      fetched <- maybeRoot match {
+                  case None =>
+                    // kickstart empty root
+                    rootsStore.recordRoot(History.emptyRootHash).map(_ => History.emptyRootHash)
+                  case Some(root: Blake2b256Hash) => Applicative[F].pure(root)
+                }
+    } yield fetched
+
+  def history(root: Blake2b256Hash): F[Blake2b256Hash] =
+    for {
+      maybeRoot <- rootsStore.validateRoot(root)
```

Comment:
this doesn't read well. 

I would change validateRoot to `isValidRoot : Blake2b256Hash -> F[Boolean]` and then the code will read better

Reply:
for now I prefer what Pawel proposed.